### PR TITLE
Loose the hashing computing without include core version

### LINKED ISSUE #6214: Loosely make hashing without include full aiida-core version
As mentioned in https://github.com/aiidateam/aiida-pseudo/issues/167#issuecomment-1809853456, the calcjob hashing objects include aiida-core version, which means if the calcjob is run with a different aiida-core the hash will change and not trigger caching. 
The `aiida-core` version was added to create hashing to avoid false positives using caching unexpectedly. 

The caching become a more popular feature so we can considered loose the condition a bit. Plus with https://github.com/aiidateam/aiida-core/pull/6156, it is easy to see if the job is run from a cached source.

### PR DESCRIPTION
**Pull request by unkcpz**
fixes #6214 

I remove the redundant version and I still keep the major version (TBD), maybe we can remove the core version.

### CODE REVIEW

**Comment by sphuber** [comment]
I think the idea is definitely interesting and we probably want to apply it, however, I think we need a bit more time to think about any issues related to backwards compatibility. We might want to add a database migration to update all existing nodes. After internal discussion, we slated this most likely for the release following v2.5.

**Review by danielhollas — CHANGES_REQUESTED** [review]

File: aiida/plugins/utils.py [hunk 1]
```diff
@@ -73,9 +73,11 @@ def get_version_info(self, plugin: str | type) -> dict[t.Any, dict[t.Any, t.Any]
         if plugin in self._cache:
             return self._cache[plugin]
 
+        # The major version of `aiida-core` is for now the only version information for hashing calculation.
+        major_version_core = version_core.split('.')[0]
         self._cache[plugin] = {
             KEY_VERSION_ROOT: {
-                KEY_VERSION_CORE: version_core,
+                KEY_VERSION_CORE: major_version_core,
```
[the remark this answers is not in the document]
Comment (unkcpz) [reply aiida/plugins/utils.py]: Yep. I define a `get_major_version` and use it to fix failed tests.

File: tests/engine/processes/calcjobs/test_calc_job.py [hunk 2]
```diff
@@ -522,6 +522,23 @@ def test_get_hash(self, get_calcjob_builder):
         _, node = launch.run_get_node(builder)
         assert node.base.extras.get(node.base.caching._HASH_EXTRA_KEY) == node.get_hash()  # pylint: disable=protected-access
 
+    def test_get_objects_to_hash(self, get_calcjob_builder):
```
[the remark this answers is not in the document]
Comment (unkcpz) [reply tests/engine/processes/calcjobs/test_calc_job.py]: Yes, changed.

File: docs/source/reference/command_line.rst [hunk 3]
```diff
@@ -573,21 +573,6 @@ Below is a list with all available subcommands.
       version    Print the current version of the storage schema.
 
 
-.. _reference:command-line:verdi-tui:
```
Comment (danielhollas) [inline docs/source/reference/command_line.rst:576]: This should be put back,

File: tests/engine/processes/calcjobs/test_calc_job.py [hunk 4]
```diff
@@ -532,6 +532,23 @@ def test_get_hash(self, get_calcjob_builder):
...
+    def test_get_objects_to_hash_core_version(self, get_calcjob_builder):
+        """Test that :meth:`aiida.orm.CalcJobNode._get_objects_to_hash` returns the expected objects."""
+        from importlib import import_module
+
+        builder = get_calcjob_builder()
+        _, node = launch.run_get_node(builder)
+        objects = node._get_objects_to_hash()
+        core_version = import_module(node.__module__.split('.', maxsplit=1)[0]).__version__
+
+        assert core_version not in objects
```
Comment (danielhollas) [inline tests/engine/processes/calcjobs/test_calc_job.py:544]: Would this assertion fail before the change?? Shouldn't it be look at `obj['version']['core']` or something like it?

**Review by danielhollas — COMMENTED** [review]
As @sphuber mentioned above, this probably needs a migration. Some comments but I do not know the caching system to do a proper review.

**Comment by unkcpz** [comment]
Hi @sphuber, I rebase this PR. You can give it a look and we can then find time to discuss. 
I think it is this https://github.com/aiidateam/aiida-core/pull/6000 PR include the migration required. We should do it separately or in this PR?

**Comment by unkcpz** [comment]
I start to need this change since I update the `aiida-core` for my production environment and nothing is now cached. Any idea on how to proceed with it? @sphuber

**Comment by sphuber** [comment]
> I start to need this change since I update the `aiida-core` for my production environment and nothing is now cached. Any idea on how to proceed with it? @sphuber

There is no problem with the implementation, but to me it is not clear whether this is the right solution and whether it doesn't introduce problems. For starters, if the version is a problem, why keep the major version. If the real thing that matters in caching is the hash of a node, and the hash is simply defined by its content, then if the content doesn't change, the hash also shouldn't change, even if the `aiida-core` version that created it changes.

Furthermore, even if we change the core version to just include the major version, the plugin version is still there. That might also change often and you would still have the same problem. This would now push it down to plugins to update their `Data` plugins to manually exclude that attribute from `_get_objects_to_hash`. If we think the version is not needed in the hash and really is just making caching useless in many cases, shouldn't we just get rid of it entirely?

And if the answer is yes, I don't think the current solution is the right one. Why change the actual versions that are stored (or even remove them)? This way we are losing valuable information. When we can simply modify `NodeCaching._get_objects_to_hash` to exclude this key. This would have the same effect, without any information loss.

Finally, there is the question whether we need a migration if we go through with this. We already changed hashing in v2.4 and added a migration for it. This was the 1st migration since v2.0, and we might have to add it once again in the next release and drop the hashes once again. Not saying we shouldn't do it therefore, but it _is_ something to keep in mind, as it will make it more difficult for users to switch between this new version and older ones.

So I think some discussion is still required before we can go on with this. Thoughts?

**Comment by sphuber** [comment]
Before I forget, another thing we might want to change while we are chaning the hashing algorithm: it would be useful if `_get_objects_to_hash` would return a dictionary instead of a list. Sometimes to debug hashing it is useful to look at the value returned by this method, but it being a list makes it difficult to identify what each value represents. If we turn it into a dictionary, that would make it easier to identify each part

File: src/aiida/orm/nodes/process/calculation/calcjob.py [hunk 5]
```diff
@@ -135,6 +132,7 @@ def _hash_ignored_attributes(cls) -> Tuple[str, ...]:  # noqa: N805
             'priority',
             'max_wallclock_seconds',
             'max_memory_kb',
+            'version',  # ?? How I exclude the version.core ??
```
Comment (unkcpz) [inline src/aiida/orm/nodes/process/calculation/calcjob.py:135]: @sphuber 
First, the attributes like `queue_name` is in `last_job_info` which is in `_node._updatable_attributes` and already be excluded, so the is no point to keep it there.
Second, it now only support exclude by key, and for the nest dictionary the login in `CalcJobNodeCaching` does not yet try to go level by level to exclude things. 

In order to make this work, I think in the `_hash_ignored_attributes` here it can specify `version.core` and in the NodeCaching class it will use the format to exclude precisely the version.core to have attribute object from: `{version: {'core': --, 'plugin': --}}` to `{'version': {'plugin': --}}`.

What you think?

Comment (sphuber) [reply src/aiida/orm/nodes/process/calculation/calcjob.py:135]: You _could_ change `_hash_ignored_attributes` and the implementation that consumes it to support nested keys, but you could also update `CalcJobNodeCaching._get_objects_to_hash()` to just call the `super` and then from the returned list, get the element that corresponds to the dictionary of attributes and then remove the 'version.core' key.

I am still not a 100% clear about the hypothetical case of the plugin being updated needing the cache to be invalidated (by the hash being different). Wouldn't the same reasoning also  not apply to `aiida-core`? What if the `CalcJob` base class changes some small (but significant) part of the implementation? Wouldn't that be the same scenario? So why keep the plugin version but drop the core version in that case? And if the changes could happen due to a fix of a bug (which could be released with a patch version of `aiida-core`) then just checking the major or even minor version _also_ won't protect against this hypothetical case.

So it seems we are back to square 1. Yes, keeping the core and plugin versions in the hash might be the safest, but how common are the cases that it is trying to protect against? If they are rare enough, should we really invalidate hashes with each (patch) version update, strongly reducing the effectiveness of the caching mechanism?

**Comment by danielhollas** [comment]
> I am still not a 100% clear about the hypothetical case of the plugin being updated needing the cache to be invalidated (by the hash being different). Wouldn't the same reasoning also not apply to aiida-core? What if the CalcJob base class changes some small (but significant) part of the implementation? Wouldn't that be the same scenario? So why keep the plugin version but drop the core version in that case? And if the changes could happen due to a fix of a bug (which could be released with a patch version of aiida-core) then just checking the major or even minor version also won't protect against this hypothetical case.
> So it seems we are back to square 1. Yes, keeping the core and plugin versions in the hash might be the safest, but how common are the cases that it is trying to protect against? If they are rare enough, should we really invalidate hashes with each (patch) version update, strongly reducing the effectiveness of the caching mechanism?

I have the same question as @sphuber. Ignoring aiida-core version but invalidating cache upon plugin update seems very strange. Especially since plugins might undergo more rapid development. It _does_ make sense in a sense that we have tight control over what we do in aiida-core, and if we did make some breaking change, we can always drop the hashes via DB migration. This would seem more difficult to handle for plugins (not to mention that plugin developers might not be aware of caching subtleties).

>  In the PR, I didn't remove the plugin version from caching, and I think at the moment we'd better still keep it there. Mentioned by @GeigerJ2, it can be false positive caching that happened if the plugin change the implementation and the calcjob give the different outputs. But inside the workchain, the cached calcjob node is taken and fail the workchain.
I remember I already did encountered the same issue during the plugin development, so even when plugin version exist it can not help to solve this issue during development.

@unkcpz @GeigerJ2 It would be very useful if you could write down couple such scenarios in detail. What are the exact conditions for the breakage? What would actually happen when there is a breaking change? Would the whole workflow crash or silently do a wrong thing? Based on that I wonder if we could have some mitigation strategies in aiida-core to prevent these issues.

**Comment by unkcpz** [comment]
> @unkcpz @GeigerJ2 It would be very useful if you could write down couple such scenarios in detail. What are the exact conditions for the breakage? 

The example I did encounter is I run a workchain of `CALCJOB_A` -> Dict_FORMAT_A -> `CALCJOB_B`. The calcjob_a will output a dict and I use the value from the dict as the input for calcjob_b (there can be some calcfunction in the middle for provenance, but doesn't matter for the moment). 

- I run the workchain and `CALCJOB_A` get cached but for some reason the `CALCJOB_B` not successfully done therefore will not be the valid cached source. 
- Let's assume now I change the output format of `CALCJOB_A` so the format it is different with containing more items, and the newly included items is used for `CALCJOB_B`.
- I rerun the workchain, I expect that the `CALCJOB_A` runs and generate new format but it actually take the old calcjob_a from cache and give the old format.

**Comment by sphuber** [comment]
I see, that would indeed fail. There is a deeper problem here. The real cause for the discrepancy in this scenario is the `Parser`. Although this _can_ (and for now almost always practically is) part of the same plugin package as the `CalcJob`, it doesn't have to be. But I admit that in practice this is unlikely as usually the calcjob and parser live in the same package.

Still, in this situation, one could "invalidate" the original `CALCJOB_A` node so it won't be used for the cache anymore. How common are these scenarios? Would it not be more reasonable to request users invalidate these old nodes if the plugins have changed significantly, rather than making caching practically ineffective by default across the board?

**Comment by danielhollas** [comment]
> Still, in this situation, one could "invalidate" the original CALCJOB_A node so it won't be used for the cache anymore. How common are these scenarios? Would it not be more reasonable to request users invalidate these old nodes if the plugins have changed significantly, rather than making caching practically ineffective by default across the board?

This sounds reasonable as long as this is thoroughly documented? How does one currently invalidate a cache for a single node? Is it possible to do through CLI? If not, we should make it as easy as possible. One disadvantage I see that the user might not be aware that the caching was the reason for the failure? Is there anything that could be done to make that more apparent?

(this is just thinking out loud. I haven't used caching myself yet so perhaps these are non-issues)

**Comment by sphuber** [comment]
> How does one currently invalidate a cache for a single node?

As @unkcpz said, there isn't a CLI endpoint yet but he has a PoC. In the Python API, you literally just do `load_node().is_valid_cache = False`. Once the CLI end point is there, I think it should be relatively easy to control for users.

**Comment by danielhollas** [comment]
> If user try to enable the caching, we pop up a warning with links sort of like that "SHOULD READ BEFORE USING CACHING".

I like that idea, we can link to the "Limitations and guidelines" section

> https://aiida.readthedocs.io/projects/aiida-core/en/latest/topics/provenance/caching.html#limitations-and-guidelines

Note that this section needs to be updated if we go forward with this PR.

> The real cause for the discrepancy in this scenario is the Parser. Although this can (and for now almost always practically is) part of the same plugin package as the CalcJob, it doesn't have to be. But I admit that in practice this is unlikely as usually the calcjob and parser live in the same package.

I don't know the CalcJob internals well, but I wonder if there's anything currently not included in the hash that would decrease the chance of a bad cache hit (i.e. something more fine-grained than the plugin version?).

**Comment by unkcpz** [comment]
> I don't know the CalcJob internals well, but I wonder if there's anything currently not included in the hash that would decrease the chance of a bad cache hit (i.e. something more fine-grained than the plugin version?).

I don't think so, you can take a look (better to be a calcjob submit to remote using non-direct scheduler which may contains more attributes) by checking the `node.base.attributes`, those items exclude the `node._updatable_attributes` are used for the hashing computing.

**Comment by sphuber** [comment]
Do you want to update this PR @unkcpz and remove the version information entirely from the hash calculation? We have given ample opportunity to discuss and I think we are converging on the conclusion that it makes caching ineffective and we are willing to take the slight increase in risk of false positives.

I would like to include these changes in the upcoming release. I just opened #6321 which fixes a bug in caching. It would be good to include all these changes in one go with a database migration that drops all hashes.

**Comment by danielhollas** [comment]
Would be great to have this, I've just started using caching and it is perfect for our use case. 

> How common are these scenarios? Would it not be more reasonable to request users invalidate these old nodes if the plugins have changed significantly, rather than making caching practically ineffective by default across the board?

I have to say that I am still uncomfortable in passing this burden on users, and would like us to think harder to make these cases less probable since caching bugs are notoriously hard to debug.

The scenario pointed by @unkcpz could be traced to the change in the parse. I am wondering if we could, instead of just copying the output nodes, we could run the current parser on top of the finished calculation? 

Another scenario would be if the plugin changed the implementation of the function that prepares the input files for the QM program (prepare_submission?). I could imagine e.g. plugin changing some default parameters of the ab initio calculations. Is this something we could capture somehow, e.g. by running the function and comparing the input files with the ones from the cached nodes? 

Another, somewhat orthogonal question -- in aiida-core we can do a DB migration and drop the hashes if the implementation changes. Do plugin authors have any possibility of invalidating the cache in a similar manner, if the plugin version is no longer part of the cache?

**Comment by sphuber** [comment]
> The scenario pointed by @unkcpz could be traced to the change in the parse. I am wondering if we could, instead of just copying the output nodes, we could run the current parser on top of the finished calculation?

This is unfortunately not really an option, because the parser could rely on files that are retrieved with the `retrieve_temporary_list` but these files are not stored in the `retrieved` node and so cannot be provided to the parser.

> Another scenario would be if the plugin changed the implementation of the function that prepares the input files for the QM program (prepare_submission?). I could imagine e.g. plugin changing some default parameters of the ab initio calculations. Is this something we could capture somehow, e.g. by running the function and comparing the input files with the ones from the cached nodes?

This would become very slow though. The whole premise of the caching mechanism is that to match nodes, a query is performed based on their hash, which is computed just once when the node is stored. With your suggestion, you would now have to find all nodes that match the hash, and then actually one-by-one start running the `prepare_for_submission` and then manually checking all input files, comparing them. This would be a huge performance hit. And then we are still not a 100% sure that we capture everything, because just checking the input files generated by `prepare_for_submission` is not everything that determines the conditions for a calculation. So I would definitely not go in this direction.

> Another, somewhat orthogonal question -- in aiida-core we can do a DB migration and drop the hashes if the implementation changes. Do plugin authors have any possibility of invalidating the cache in a similar manner, if the plugin version is no longer part of the cache?

No they cannot, only `aiida-core` can add migrations. Users would have to essentially perform a query and invalidate those nodes, e.g. something like
```
for node, in QueryBuilder().append(PwCalculation).iterall():
    node.is_valid_cache = False
```

**Comment by unkcpz** [comment]
Hi @sphuber, I am going to finish this one today. So in the end, we agree on remove all versions from caching correct? If so this PR would be then a tiny one. 
Also mentioned in https://github.com/aiidateam/aiida-core/pull/6215#issuecomment-1961608709, do we require the change on the prompt when activating the caching?

**Comment by sphuber** [comment]
> I am going to finish this one today. So in the end, we agree on remove all versions from caching correct?

I did suggest to go on with removing all versions from the hash, but then @danielhollas voiced concerns. I responded to his suggestions, explaining why I think they are not feasible. Not sure if those counter comments were satisfactory.

> Also mentioned in [#6215 (comment)](https://github.com/aiidateam/aiida-core/pull/6215#issuecomment-1961608709), do we require the change on the prompt when activating the caching?

Not sure to be honest. It would require a special exception in `Config.set_option` which feels a bit dirty. I think this should simply be part of the caching documentation.

**Comment by unkcpz** [comment]
> tl;dr @unkcpz please go ahead with the implementation in this PR.

👍 Will do it. 

> In most cases, the prepare_submission would be only called on a single node that matched

I think I sort of agree with @danielhollas and think this would bring a bit robust to the caching mechanism. In the end it may not have too many cached nodes found.

**Comment by unkcpz** [comment]
Hi @sphuber, I think this is ready for the final review.

**Comment by sphuber** [comment]
Thanks @unkcpz . I gave the changes a quick look and they look ok to me. I would just first want to make sure that what I suggested in [this comment](https://github.com/aiidateam/aiida-core/pull/5998#issuecomment-2004788531) would provide the necessary control to plugin developers to essentially reset the cache if the implementation of a `Data` or `CalcJob` plugin changes significantly. I will try to make a PR tonight/tomorrow and then @danielhollas and yourself can maybe take a look at it. If that seems sufficient, we can merge both changes.

**Comment by sphuber** [comment]
Thanks @unkcpz . I gave the changes a quick look and they look ok to me. I would just first want to make sure that what I suggested in [this comment](https://github.com/aiidateam/aiida-core/pull/5998#issuecomment-2004788531) would provide the necessary control to plugin developers to essentially reset the cache if the implementation of a `Data` or `CalcJob` plugin changes significantly.

I just opened #6328 as a draft PR that attempts an implementation, but as I describe in detail there, it seems not so easy after all...

File: tests/orm/nodes/test_node.py [hunk 6]
```diff
@@ -981,15 +981,25 @@ def test_store_from_cache(self, tmp_path):
...
-    def test_hashing_errors(self, aiida_caplog):
+    def test_hashing_errors(self, aiida_caplog, monkeypatch):
         """Tests that ``get_hash`` fails in an expected manner."""
+        # monkeypatch `get_objects_to_hash` to raise an fake error
+        from aiida.orm.nodes.caching import NodeCaching
+
         node = Data().store()
-        node.__module__ = 'unknown'  # this will inhibit package version determination
+
+        # monkeypatch `get_objects_to_hash` to raise an fake error
```
Comment (danielhollas) [inline tests/orm/nodes/test_node.py:991]:
```suggestion
        # monkeypatch `get_objects_to_hash` to raise a fake error
```

File: tests/engine/processes/calcjobs/test_calc_job.py [hunk 7]
```diff
@@ -509,6 +509,15 @@ def test_get_hash(self, get_calcjob_builder):
         _, node = launch.run_get_node(builder)
         assert node.base.extras.get(node.base.caching._HASH_EXTRA_KEY) == node.base.caching.get_hash()
 
+    def test_get_objects_attributes(self, get_calcjob_builder):
+        """Test that :meth:`aiida.orm.CalcJobNode._get_objects_to_hash` returns the expected objects."""
+        builder = get_calcjob_builder()
+        _, node = launch.run_get_node(builder)
+        objects = node.base.caching.get_objects_to_hash()  # pylint: disable=protected-access
+
+        assert 'version' not in objects
```
Comment (danielhollas) [inline tests/engine/processes/calcjobs/test_calc_job.py:518]: This is a bit of a proxy test. Might be nice to have another test that would specifically check that the hash is independent of aiida-core version.

**Review by danielhollas — APPROVED** [review]
LGTM! (with one suggestion for an extra test). 
Might be nice to get this landed first and integrate it to #6328 so that we can test everything end-to-end.

File: tests/engine/processes/calcjobs/test_calc_job.py (hunk 7 above)
Comment (sphuber) [reply tests/engine/processes/calcjobs/test_calc_job.py:521]: Done

**Review by sphuber — APPROVED** [review]
Thanks to you both @unkcpz and @danielhollas

**Comment by unkcpz** [comment]
Thanks for taking care of this @sphuber and sorry for not active aiida stuff this month.
